### PR TITLE
Add support for PVR Providers (HTSPv38+)

### PR DESCRIPTION
**Pull request by ksooo**
Requires a decent tvheadend build, at least [267aef151ec30fa9c5469500100ab7f59092d39a](https://github.com/tvheadend/tvheadend/commit/267aef151ec30fa9c5469500100ab7f59092d39a) (merged Aug 23rd, 2024)

### CODE REVIEW

File: src/Tvheadend.cpp [hunk 1]
```diff
@@ -223,6 +224,50 @@ bool CTvheadend::HasStreamingProfile(const std::string& streamingProfile) const
...
+    {
+      kodi::addon::PVRProvider provider;
+      provider.SetUniqueId(entry.second.GetId());
+      provider.SetName(entry.second.GetName());
+      //SetType
+      //SetIconPath
+      //SetCountries
+      //SetLanguages
+
+      providers.emplace_back(std::move(provider));
```
Comment (Jalle19) [inline src/Tvheadend.cpp:254]: Why not `std::move()` directly to `results`?

File: src/Tvheadend.cpp [hunk 2]
```diff
@@ -2256,10 +2348,29 @@ void CTvheadend::ParseChannelDelete(htsmsg_t* msg)
...
+  const auto it = m_channels.find(u32);
+  if (it != m_channels.cend())
+    providerUid = (*it).second.GetProviderUid();
+
+  /* Erase channel */
   m_channels.erase(u32);
   m_channelTuningPredictor.RemoveChannel(u32);
   TriggerChannelUpdate();
+
+  if (providerUid > 0)
```
Comment (Jalle19) [inline src/Tvheadend.cpp]: Should this be `providerUid != PVR_PROVIDER_INVALID_UID`?

File: src/Tvheadend.cpp [hunk 3]
```diff
@@ -2224,6 +2291,31 @@ void CTvheadend::ParseChannelAddOrUpdate(htsmsg_t* msg, bool bAdd)
       /* CAID */
       if (caid == 0)
         htsmsg_get_u32(&f->hmf_msg, "caid", &caid);
+
+      /* Service provider */
+      str = htsmsg_get_str(&f->hmf_msg, "providername");
+      if (str)
+      {
+        const uint32_t uid{utilities::hash_str_uint32(str)};
```
Comment (Jalle19) [inline src/Tvheadend.cpp]: Empty provider names should probably be ignored?

File: src/Tvheadend.cpp (hunk 1 above)
Comment (ksooo) [reply src/Tvheadend.cpp:254]: `results` is not a simple container. `Add` actually calls back to Kodi to immediately deliver the result. We use the concept "collect results mutex ptrotected, then release mutex and deliver results to Kodi" everywhere in the add-on, also for channels, groups, ...

File: src/Tvheadend.cpp (hunk 2 above)
Comment (ksooo) [reply src/Tvheadend.cpp]: Yep. Thanks.

File: src/Tvheadend.cpp (hunk 3 above)
Comment (ksooo) [reply src/Tvheadend.cpp]: Yep. Thanks.

File: src/Tvheadend.cpp (hunk 2 above)
Comment (ksooo) [reply src/Tvheadend.cpp]: Wait! PVR_PROVIDER_INVALID_UID is signed and defined as -1. providerUid is unsigned.

Comment (ksooo) [reply src/Tvheadend.cpp]: Will change type accordingly.

Comment (ksooo) [reply src/Tvheadend.cpp]: done

File: src/Tvheadend.cpp (hunk 3 above)
Comment (ksooo) [reply src/Tvheadend.cpp]: done

File: src/Tvheadend.cpp (hunk 1 above)
Comment (Jalle19) [reply src/Tvheadend.cpp:254]: Right :+1:

File: src/Tvheadend.cpp [hunk 4]
```diff
@@ -2256,10 +2344,29 @@ void CTvheadend::ParseChannelDelete(htsmsg_t* msg)
   }
   Logger::Log(LogLevel::LEVEL_DEBUG, "delete channel %u", u32);
 
-  /* Erase */
+  /* We need to check and update providers. May be the deleted channel is the last channel with this provider */
+  int32_t providerUid{0};
```
Comment (Jalle19) [inline src/Tvheadend.cpp]: Should this too be `PVR_PROVIDER_INVALID_UID`?

Comment (ksooo) [reply src/Tvheadend.cpp]: Yes. I will fix that.

Comment (ksooo) [reply src/Tvheadend.cpp]: Done.

**Review by Jalle19 — APPROVED** [review]
Nice feature, thanks!

**Pull request merged**